# Klasser
Under kursens gång har vi pratat om enkla datatyper som *heltal*, *strängar* med flera, och sammansatta datatyper såsom *listor*, *dictionaries* och *set*. Vi har till och med skapat egna komplicerade datastrukturer, såsom vår tabellmodell.

I synnerhet dictionaries har visat sig vara väldigt passande för att modellera komplicerade data med många attribut, exempelvis vårt ursprungliga exempel, med en katt. 

| Antal ben | Vikt | Namn | Chippad? |
| --- | --- | --- | --- |
| 4 | 2.1 | "Heidi" | sant |

I exemplet ovan kan vi välja att modellera katten som en sammansatt datatyp, ett dictionary med lite olika nyckel-värdepar:

In [1]:
katt = {
    'antal ben': 4,
    'vikt': 2.1,
    'namn': 'Heidi',
    'chippad?': True
}

Men faktum kvarstår att detta inte är en katt, utan ett dictionary. Det är en grov abstraktion av en katt, och vi kan dessutom inte avkräva någon annan användare som får detta dictionary att de respekterar att det är en katt vi modellerat, och inte exempelvis lägger till ett nytt attribut:

In [2]:
katt['drivmedel'] = 'diesel'
# katt.update({'drivmedel': 'diesel'}) # Alternatively

I slutändan är de så kallade *primitiva datatyperna* (heltal, flyttal, boolean) otillräckliga. Inte heller de sammansatta datatyperna (lista, dict, set, o.s.v.) ger oss ett särskilt kraftfullt sätt att modellera data. 

Vi söker därför medlen att skapa ännu mer komplexa beskrivningar av verkligheten, och mer potenta abstraktioner. Nyckeln är i vårt fall *objekt-orienterad programmering*.

## Objekt-orienterad programmering (OOP)
Objekt-orienterad programmering är ett så kallat programmeringsparadigm. Det innebär att det är ett konventionaliserat sätt att strukturera upp sin kod och sina program, och att det därmed finns andra alternativ. 

Centralt för OOP är två begrepp, *klass* och *objekt*. Dessa är direkta generaliseringar av *datatyp* och *instans* som vi nämnde i början av kursen.

### Klass och objekt
En klass i Python är en abstraktion av någon form av data eller verklig företeelse. En klass har attribut med vissa värden, men också egna inbyggda funktioner. I enkla termer är en klass en *mall* för att skapa enskilda *instanser* av klassen. 

Vi har en definition av vad det innebär att vara en katt eller säg, ett däggdjur. Emellertid så är katten *Heidi* eller kon som producerat frukostmjölken två specifika exempel på en katt eller ett däggdjur. Dessa instanser kallas i Pythons fall för *objekt*.

En klass skrivs som en mall för vad det är man vill beskriva. Inuti klassens *scope* så kan man definiera funktioner och variabler, som endast kommer vara tillgängliga inuti klassen, precis som för enskilda funktioner. 

In [10]:
class Cat:

    # Define variables common to all cats
    number_of_legs = 4
    weight = 2.1
    name = 'Heidi'
    is_chipped = True

    def say_meoew():
        print("Meow!")

Vi kan komma åt funktioner och variabler genom den så kallade *punkt-notationen*. 

In [14]:
print(Cat.is_chipped)
print(Cat.name)
print(Cat.number_of_legs)
Cat.say_meoew()

True
Heidi
4
Meow!


Vi har nu skapat en klass. Detta är endast en mall, och är i sig meningslös. Variablerna vi skapat är fixa, och gemensamma för alla katter. Men alla katter heter inte samma sak, väger inte samma, och vissa säger inte ens *mjau*.

Vi skall därför anpassa vår klass så att vi kan skapa, individuella katter, katt-objekt.

För att faktiskt skapa ett katt-objekt, måste vi så kallat *initialisera* klassen till ett objekt. Detta görs genom att definiera en specialfunktion som alltid heter *__init__* (med två understreck på varje sida).

Den tar ett argument som kallas *self*. Man kan även döpa variabeln till exempelvis *this*. Poängen är att den variabeln syftar på varje ny instans av en katt, och inte alla katter. 

In [22]:
class Cat:

    # All cats have 4 legs (from the start at least...)
    number_of_legs = 4

    # When creating new, individual cases of cats, adapt these features
    def __init__(self, weight, name, chipped, colour):
        
        self.weight = weight
        self.name = name
        self.colour = colour
        self.is_chipped = chipped

Klasser brukar av tradition skrivas med stor bokstav, och variabler och objekt med liten bokstav.

En nytt objekt av en klass skapas av att man skriver motsvarande

```python

# A variable called obj, pointing to a new object of a class called class_name
obj = Class_name(...)
```

In [18]:
cat_obj = Cat(4.1, 'Heidi', True, "orange")

print(cat_obj.number_of_legs)
print(cat_obj.name)


print("A new cat " + cat_obj.name + " is born, weighing " + str(cat_obj.weight) + " kg, having " + cat_obj.colour + " fur and " + str(cat_obj.number_of_legs) + " legs.")

4
Heidi
A new cat Heidi is born, weighing 4.1 kg, having orange fur and 4 legs.


Men vi kan med enkelhet skapa fler katter, som inte är samma som innan.

In [19]:
another_cat = Cat(2.1, 'Smoky', False, "black")
new_cat = Cat(weight=43, name='Chonky', chipped=True, colour="green")

print(another_cat.name)
print(new_cat.name)

Smoky
Chonky


Observera dock att endast egenskaperna vi definierat genom att använda `self` är individuella, som 

```python
self.colour = "orange"
```

Attributet som är definierat i själva klassen, `number_of_legs` är ett klass-attribut, och gemensamt för alla katt-objekt:

In [21]:

print(cat_obj.number_of_legs)
print(new_cat.number_of_legs)
print(another_cat.number_of_legs)

4
4
4


## Ett annat exempel
Vi tar nu ett exempel som utnyttjar inbyggda funktioner litet mer hos klasser och objekt. Betrakta en bil, dessa har en topphastighet och en acceleration, som är gemensam för alla bilar av ett nytt märke (åtminstone till en början). 

En annan egenskap hos en bil är den hastighet den har vid ett specifikt tillfälle. Denna är dock unik för en individuell bil, alla bilar kör ju inte alltid i samma hastighet. 

Till att börja med står en bil stilla. Det betyder att alla bil-objekt initialiseras med samma starthastighet, 0 (`self.speed = 0`).

In [28]:
class Car:
    top_speed = 150 #km/h
    acceleration = 100 #km/h^2

    def __init__(self):

        self.speed = 0

    def accelerate(self, time):

        self.speed = self.acceleration * time


Medlemsfunktionen `accelerate` tar ett argument `self`, som syftar på att den tillhör objektet och inte klassen, samt `time` som är hur länge en bil accelererar.

Vi skapar en bil och låter den accelerera i en halvtimme:

In [32]:
c = Car()

# Inspect speed at creation
print(c.speed)

# Accelerate
c.accelerate(0.5)

# Inspect speed after acceleration
print(c.speed)

0
50.0


Ganska långsam acceleration. Men om vi låter den accelera ännu längre, vad händer då?

In [33]:
print(c.speed)

# Accelerate for a while longer
c.accelerate(2.5)

print(c.speed)


50.0
250.0


Aj då... Det betyder att bilen nu har en hastighet på 250 km/h, vilket övergår hastigheten väsentligt. Vi får fixa till vår definition:

In [34]:
class Car:
    top_speed = 150 #km/h
    acceleration = 100 #km/h^2

    def __init__(self):

        self.speed = 0

    def accelerate(self, time):

        self.speed = self.acceleration * time

        # If the current speed would be higher than
        # the top speed, limit the speed to the top_speed
        if self.speed > self.top_speed:

            self.speed = self.top_speed

In [36]:
c = Car() # Create a new object

print(c.speed)

c.accelerate(5) # Accelerate for 5 hours

print(c.speed)

0
150


## Fler exempel
Faktum är att ni vid flera tillfällen stött på klasser och objekt. Här är en kort lista:

- Listor (med medlemsfunktioner såsom `append` och `pop`)
- Dictionaries (med medlemsfunktioner som `update`, och där varje nyckel-värdepar är attribut)
- Strängar (med medlemsfunktioner som `replace` och `split`)

Dessa är dock så centrala att de fått lite extra fin syntax. Andra exempel är
- `pd.DataFrame`, en tabellklass med flera hjälpsamma funktioner
- `Counter`, som räknar antal förekomster i en lista

## Mer om `self` och jämförelse
Vad är det här mystiska argumentet `self` som vi skickar runt? 

I enklaste ordalag så är det individuella objektet själv, en referens till just det objektet av en klass, och inget annat. Vi kan titta litet noggrannare genom att skriva ut resultatet när vi initialiserar/skapar ett objekt:

In [37]:
class Person:
    
    def __init__(self, name):
        self.name = name

        # When created, prints the self variable value
        print(self)

In [40]:
Person("Elton John")

Detta producerar ett resultat `<__main__.Person object at ...>` där det sista är en kod i hexadecimalt format. Detta är exakt var det nuvarande objektet sparas i minnet, och är unikt för varje nytt, skapat objekt. Inte ens om vi skapar ett nytt objekt med samma egenskaper (namn, i det här fallet), så kommer objektet att vara likadant i minnet.

In [39]:
Person("Elton John")

I princip är detta definitionen av instans, en enskild individ skapad ur samma mall (eller klass). En konsekvens av detta är att de inte är jämförbara. Likhet mellan objekt är inte definierad, för de är alla olika:

In [41]:
print(Person("Elton John") == Person("Elton John"))

False


I stället brukar man definiera en "equals"-funktion hos klasser, som säger om två objekt av samma klass är likadana. Vi kan då titta på vilka objekt som skickas in i respektive inargument:

In [47]:
class Person:
    
    def __init__(self, name):
        self.name = name

    def equals(self, other):

        print("Self: ", self)
        print("Other: ", other)


        if self.name == other.name:
            return True
        else:
            return False

In [51]:
p1 = Person("Elton John")
p2 = Person("David Bowie")
p3 = Person("Elton John")


print("Are they equal?", p1.equals(p2))
print("Are they equal?", p1.equals(p3))


Self:  <__main__.Person object at 0x7f4e065c2fb0>
Other:  <__main__.Person object at 0x7f4e1c1eee00>
Are they equal? False
Self:  <__main__.Person object at 0x7f4e065c2fb0>
Other:  <__main__.Person object at 0x7f4e065c2da0>
Are they equal? True
